07_VirtualFileSystem/02_VirtualFileSystem.md

```diff
@@ -0,0 +1,352 @@
+* Return a file descriptor or an error if needed. 
+
+The function header for our open function will be: 
+
+```c 
+int open(const char *filename, int flags);
+```
+
+The `flags` parameter will tell how the file will be opened, there are many flags, and the three below should be mutually exclusive:
+
+* O_RDONLY it opens a file in read only mode
+* O_RDWR it opens a file for reading and writing
+* O_WRONLY it opens a file only for writing.
+
+The flags value is a bitwise operator, and there are other possible values to be used, but for our purpose we will just use the three above. 
+
+The return value of the function is the file descriptor id. We have already seen how to parse a path and get the mountpoint id if it is available. But what about the file descriptor and it's id? What is it? File descriptors represents a file that has been opened by the VFS, and contain information on how to access it (i.e. mountpoint_id), the filename, the various pointers to keep track of current read/write positions, eventual locks, etc. So before proceed let's outline a very simple file descriptor struct: 
+
+```c
+struct {
+    int fs_file_id;
+    int mountpoint_id;
+    char *filename;
+    int buf_read_pos;
+    int buf_write_pos;
+    char *file_buffer;
+} file_descriptor_t
+```
+
+Where the `mountpoint_id` fields is the id of the mounted file system that is contining the requeste file. The `fs_file_id` is the fs specific id of the fs opened file descriptor. `buf_read_pos` and `buf_write_pos` are the current positions of the buffer pointer for the read and write operations. 
+
+So once our open function has found the mountpoint for the requested file, eventually a new file descriptor item will be created and filled, and an id value returned. This id is different from the ine in the data structure, since it represent the internal fs descriptor id, while this one represent the vfs descriptor id. In our case the descriptor list is implemented again using an array, so the id returned will be the array position where the descriptor is being filled.
+
+Why "eventually" ? Having found the mountpoint id for the file doesn't mean that the file exists on that fs, the only thing that exist so far is the mountpoint, but after that the VFS can't really know if the file exists or not, it has to devolve this task to the fs driver, hence it will call the implementation of a function that open a file on that FS that will do the search and return the an error if the file doesn't exists.  
+
+But how to call the fs driver function? Earlier in this chapter when we outlined the `mountpoint_t` structure we added a field called `operations`, of type `fs_operations_t` and left it unimplemented. Now is the time to implement it, this field is going to contain the pointer to the driver functions that will be used by the vfs to open, read, write, and close files:
+
+```c
+struct fs_operations_t {
+	int (*open)(const char *path, int flags, ... );
```

Comment:
is the `...` meant to be there haha? Let's keep things simple and the bare minimum.